File: PoketDex/app/src/main/java/pt/rfsfernandes/ui/fragments/PokemonDetailsFragment.java
package pt.rfsfernandes.ui.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import pt.rfsfernandes.MyApplication;
import pt.rfsfernandes.R;
import pt.rfsfernandes.custom.adapters.PokemonMovesAdapter;
import pt.rfsfernandes.custom.adapters.view_pager.DetailsPagerAdapter;
import pt.rfsfernandes.databinding.FragmentPokemonDetailsBinding;
import pt.rfsfernandes.model.pokemon.Pokemon;
import pt.rfsfernandes.viewmodels.MainViewModel;

import static androidx.fragment.app.FragmentStatePagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT;


public class PokemonDetailsFragment extends Fragment {
  private FragmentPokemonDetailsBinding binding;
  private MainViewModel mMainViewModel;
  private MyApplication mMyApplication;
  private PokemonMovesAdapter mPokemonMovesAdapter;
  private DetailsPagerAdapter mDetailsPagerAdapter;
  private int maxPage = 1;

  public PokemonDetailsFragment() {
    // Required empty public constructor
  }


  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {
    // Inflate the layout for this fragment
    binding = FragmentPokemonDetailsBinding.inflate(inflater, container, false);
    View view = binding.getRoot();
    mMainViewModel = new ViewModelProvider(requireActivity()).get(MainViewModel.class);
    mMyApplication = (MyApplication) requireActivity().getApplication();
    setupViews();
    return view;
  }

  /**
   * Sets up views
   */
  private void setupViews() {
    mDetailsPagerAdapter = new DetailsPagerAdapter(getChildFragmentManager(), BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
    binding.viewPagerDetails.setAdapter(mDetailsPagerAdapter);
    maxPage = mDetailsPagerAdapter.getCount();
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);

    binding.imageButtonLeft.setOnClickListener(e -> {
      if (mMyApplication.isCanPlaySounds()) {
        mMyApplication.getMediaPlayerMenuSound().start();
      }
      if (binding.viewPagerDetails.getCurrentItem() == 0) {
        mMainViewModel.changePage(mDetailsPagerAdapter.getCount());
      } else {
        mMainViewModel.changePage(binding.viewPagerDetails.getCurrentItem() - 1);
      }

    });

    binding.imageButtonRight.setOnClickListener(e -> {
      if (mMyApplication.isCanPlaySounds()) {
        mMyApplication.getMediaPlayerMenuSound().start();
      }
      if (binding.viewPagerDetails.getCurrentItem() == maxPage - 1) {
        mMainViewModel.changePage(0);
      } else {
        mMainViewModel.changePage(binding.viewPagerDetails.getCurrentItem() + 1);
      }

    });

  }

  @Override
  public void onResume() {
    super.onResume();
    initViewModel();
  }

  @Override
  public void onDetach() {
    super.onDetach();
  }

  /**
   * Initiates viewModel observers
   */
  private void initViewModel() {
    mMainViewModel.getIsLoadingMutableLiveData().observe(getViewLifecycleOwner(), isLoading -> {
//      binding.content.setVisibility(isLoading ? View.GONE : View.VISIBLE);
      binding.progressBarGeneral.setVisibility(isLoading ? View.VISIBLE : View.GONE);
    });

    mMainViewModel.getPokemonMutableLiveData().observe(getViewLifecycleOwner(), pokemon -> {
      if (pokemon != null) {
//        binding.textViewPagerTitle.setText("");
        binding.textViewPokemonNumber.setText(String.format("%s%s",
            getResources().getString(R.string.no), pokemon.getId()));

      }
    });

    mMainViewModel.getDetailsPagerLiveData().observe(getViewLifecycleOwner(), page -> {
      binding.viewPagerDetails.setCurrentItem(page, true);
    });

    mMainViewModel.getDetailsTitleLiveData().observe(getViewLifecycleOwner(), title -> {
      binding.textViewPagerTitle.setText(title);
    });

  }

  /**
   * inflates dynamically linearLayoutAbilitiesContainer with each ability and refreshes the
   * moves adapter with the selected pokemon moves
   *
   * @param pokemon selected pokemon
   */
  private void fillMovesAndAbilities(Pokemon pokemon) {
//    binding.linearLayoutAbilitiesContainer.removeAllViews();
//    mPokemonMovesAdapter.refreshList(pokemon.getMoveslist());
//    for (Ability ability :
//        pokemon.getAbilitiesList()) {
//      View view = getLayoutInflater().inflate(R.layout.ability_row,
//          binding.linearLayoutAbilitiesContainer, false);
//
//      TextView textViewAbilityName =
//          view.findViewById(R.id.textViewAbilityName);
//      textViewAbilityName.setText(UtilsClass.toCamelCase(ability.getAbility().getName()));
//
//      TextView textViewAbilitySlot =
//          view.findViewById(R.id.textViewAbilitySlot);
//
//      view.findViewById(R.id.textViewAbilityHidden).setVisibility(ability.isHidden() ? View.INVISIBLE : View.VISIBLE);
//
//      textViewAbilitySlot.setText(String.format("%s %s", getString(R.string.slot_n), ability.getSlot()));
//
//      binding.linearLayoutAbilitiesContainer.addView(view);
//
//    }
  }

}

File: PoketDex/app/src/main/java/pt/rfsfernandes/pocketdex/ui/fragments/PokemonDetailsFragment.kt
package pt.rfsfernandes.pocketdex.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import pt.rfsfernandes.pocketdex.R
import pt.rfsfernandes.pocketdex.custom.adapters.view_pager.DetailsPagerAdapter
import pt.rfsfernandes.pocketdex.databinding.FragmentPokemonDetailsBinding
import pt.rfsfernandes.pocketdex.MyCustomApplication
import pt.rfsfernandes.pocketdex.model.pokemon.Pokemon
import pt.rfsfernandes.pocketdex.viewmodels.MainViewModel

class PokemonDetailsFragment : Fragment() {
    private lateinit var binding: FragmentPokemonDetailsBinding
    private val mMainViewModel: MainViewModel by sharedViewModel()
    private lateinit var mMyCustomApplication: MyCustomApplication
    private lateinit var mDetailsPagerAdapter: DetailsPagerAdapter
    private var maxPage = 1
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        binding = FragmentPokemonDetailsBinding.inflate(inflater, container, false)
        val view: View = binding.root
        mMyCustomApplication = requireActivity().application as MyCustomApplication
        setupViews()
        return view
    }

    /**
     * Sets up views
     */
    private fun setupViews() {
        mDetailsPagerAdapter = DetailsPagerAdapter(childFragmentManager,
                FragmentStatePagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT, context)
        binding.viewPagerDetails.adapter = mDetailsPagerAdapter
        maxPage = mDetailsPagerAdapter.count
        binding.viewPagerDetails.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                if (positionOffset == 0f) {
                    binding.textViewPagerTitle.text = mDetailsPagerAdapter.getPageTitle(position)
                }
            }

            override fun onPageSelected(position: Int) {
                binding.textViewPagerTitle.text = mDetailsPagerAdapter.getPageTitle(position)
            }

            override fun onPageScrollStateChanged(state: Int) {}
        })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.imageButtonLeft.setOnClickListener { e: View? ->
            if (mMyCustomApplication.canPlaySounds) {
                mMyCustomApplication.mediaPlayerMenuSound!!.start()
            }
            val position: Int = if (binding.viewPagerDetails.currentItem == 0) {
                mDetailsPagerAdapter.count
            } else {
                binding.viewPagerDetails.currentItem - 1
            }
            mMainViewModel.changePage(position, false, requireActivity())
        }
        binding.imageButtonRight.setOnClickListener { e: View? ->
            if (mMyCustomApplication.canPlaySounds) {
                mMyCustomApplication.mediaPlayerMenuSound!!.start()
            }
            val position: Int = if (binding.viewPagerDetails.currentItem == maxPage - 1) {
                0
            } else {
                binding.viewPagerDetails.currentItem + 1
            }
            mMainViewModel.changePage(position, false, requireActivity())
        }
    }

    override fun onResume() {
        super.onResume()
        initViewModel()
    }

    /**
     * Initiates viewModel observers
     */
    private fun initViewModel() {
        mMainViewModel.isLoadingMutableLiveData.observe(viewLifecycleOwner) { isLoading: Boolean ->
            binding.progressBarGeneral.visibility = if (isLoading) View.VISIBLE else View.GONE
        }
        mMainViewModel.pokemonMutableLiveData.observe(viewLifecycleOwner) { pokemon: Pokemon? ->
            if (pokemon != null) {
                binding.textViewPokemonNumber.text = String.format("%s%s",
                        resources.getString(R.string.no), pokemon.id)
            }
        }
        mMainViewModel.detailsPagerLiveData.observe(viewLifecycleOwner) { page: Int? -> binding.viewPagerDetails.setCurrentItem(page!!, true) }
    }
}